### DISCUSSION TITLE
Pattern matching for event method

### DISCUSSION THREAD

**Discussion by Jogge** [dotnet/csharplang#4906] [Language Ideas]
Often when I am using events I have to check wether the sender is a specific type like this:

```C#
DataContextChanged += (sender, _) =>
{
    if (sender is MyViewModel myViewModel)
    {
        myViewModel.DoSomeThing();
    }
};
```

Why is it not possible to do pattern matching in the signature of the method?

Example:

```C#
DataContextChanged += (sender is MyViewModel myViewModel, _) =>
{
    myViewModel.DoSomeThing();
};
```

I would expect only to run `myViewModel.DoSomeThing();` when `sender` acutually is of type `MyViewModel`. If it is not, I would expect the event to ignore this method.

**Comment by jnm2** [comment]
"Why is it not possible to do pattern matching in the signature of the method" is an easy question that I expect everyone will answer first. The most general response is that language design is never "why not," always "why, and how does this balance against the -100 points for any new feature?" More specifically to answer "why" in this case, pattern matching and lambda parameters have never been considered together from my memory of meeting notes. The specification work seems like it could be complicated, introducing for the first time the concept of a conditional lambda.

> If it is not, I would expect the event to ignore this method.

This doesn't seem like the easiest way to talk about what happens. The event probably can't feasibly be made to know whether this is a pattern-matching lambda or not. Probably the event only sees a handler delegate like any other delegate, and the pattern matching is lowered to the equivalent of an `if` statement in the lambda.

```cs
// Rough idea of lowered code
DataContextChanged += ($a$, _) =>
{
    if ($a$ is MyViewModel myViewModel)
    {
        myViewModel.DoSomeThing();
    }
    // Does an implicit `return` statement get generated if the delegate type has a non-void return?
    // Seems like a confusing idea, so probably best to only allow this for lambdas that end up being void-returning.
};
```

A similar genre that has been suggested is to allow deconstruction in lambda parameters: https://github.com/dotnet/csharplang/discussions/258. This makes me think that it might be better to start pattern matching right in place with no parameter name or `is` keyword. It's probably too subtle though; reading the code, it is easy to not notice the pattern matching and think that this is unconditional, or a cast like `foreach (MyViewModel myViewModel in objects)` would be.

```cs
DataContextChanged += (MyViewModel myViewModel, _) =>
{
    myViewModel.DoSomeThing();
};
```

**Reply by HaloFour** [reply]
I know pattern matching and overloads were proposed and there was a negative reaction to that by some of the language design team members.  I'm trying to search around for it but I've been unsuccessful so far.

One can argue that it applies less in this case given that a lambda isn't a part of the API surface of the type.  But, IIRC, one of the objections was that it mixes logic into the signature of the method/lambda, and I think that does still apply here.  As @jnm2 pointed out there's also the issue that, for a lambda that is expected to return a value, there's really no good thing for the compiler to do but to emit throwing an exception, like with an unmatched `switch` expression.  Event delegates are likely to return `void` so perhaps less of an issue.  At that point it's a question of whether such shorthand is interesting enough and applies often enough to be worth the investment.
